x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/role/BulkPutRolesResponse.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.core.security.action.role;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.DocWriteResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.List;
+
+public class BulkPutRolesResponse extends ActionResponse implements ToXContentObject {
+
+    private final List<Item> items;
+    private final boolean error;
+    private final long tookInMillis;
+
+    public static class Builder {
+
+        private final List<Item> items = new LinkedList<>();
```

Comment:
The order in the **request** body is not "stable enough" to be used in the response body here. Marshaling and unmarshaling dictionaries (by the client lib itself, or by proxies) might change the order. It works OK for the bulk index API because the request body there is not valid JSON.

I would strongly suggest we adopt the response format of the bulk api key update: https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-bulk-update-api-keys.html#security-api-bulk-update-api-keys-response-body (a list of updated role names, a list of noops, and an object of errors).

Reply:
The update api key api is a little different because the ordering of the ids array is stable and that means we can rely on it for the order of responses being in order too (within its result groups). 

My proposal is to: 
1. Order the request by changing the type from a map to a list of entries in the request: 
`{"roles": [{"role1": {descriptor_for_role1}}, ... {"roleN": {descriptor_for_roleN}]}`
(Adding a name field to the descriptor instead of a name -> descritptor mapping only for purposes of this API feels wrong, but open to discussing it)

2. Keep the response aligned with the bulk ingest API. 
I think dealing with one response item and it's status at a time, instead of all created/updated/noops/errors in groups or not sorted at all, makes for a better API. 

The alternative to this approach is to keep the map in the request and disregard the ordering and grouping of results in status buckets of created, updated, noops, errors.